src/quest/serializer.py

```diff
@@ -0,0 +1,99 @@
+from typing import Any, Dict, Protocol, TypeVar, Callable, Tuple, TypedDict
+from .utils import get_obj_name, import_object
+import inspect
+
+T = TypeVar('T')
+
+
+class TypeFactory(Protocol[T]):
+    def __call__(self, *args: Any, **kwargs: Any) -> T:
+        ...
+
+
+class SerializedData(TypedDict):
+    _ms_factory: str
+    args: Any
+    kwargs: Any
+
+
+class StepSerializer(Protocol):
+    async def serialize(self, obj: Any) -> Any:
+        ...
+
+    async def deserialize(self, data: Any) -> Any:
+        ...
+
+
+class TypeSerializer(Protocol[T]):
+    async def serialize(self, obj: T) -> Tuple[Callable[..., T], Tuple[Any, ...], Dict[str, Any]]:
+        ...
+
+
+class NoopSerializer(StepSerializer):
+    async def serialize(self, obj: Any) -> Any:
+        # object already JSON-serializable
+        return obj
+
+    async def deserialize(self, data: Any) -> Any:
+        return data
+
+
+class MasterSerializer(StepSerializer):
+    def __init__(self, type_serializers: Dict[type, TypeSerializer[Any]]):
+        self._type_serializers = type_serializers
```

Comment:
`self._type_serializers = {str(tp): ser for tp, ser in type_serializers.items()}`

This will make it easier to look up a serializer from the type stored in the SerializedData record.